crates/prune/prune/src/segments/history.rs

```diff
@@ -37,42 +37,30 @@ where
     for sharded_key in highest_sharded_keys {
         // Seek to the shard that has the key >= the given sharded key
         // TODO: optimize
-        let result = cursor
-            .seek(RawKey::new(sharded_key.clone()))?
-            .map(|(key, value)| Result::<_, DatabaseError>::Ok((key.key()?, value)))
-            .transpose()?;
-
-        // If such shard doesn't exist, skip to the next sharded key
-        if result.as_ref().map_or(true, |(key, _)| !key_matches(key, &sharded_key)) {
-            continue
-        }
-
-        // At this point, we're sure that the shard with the given sharded key exists
-        let (key, raw_blocks): (T::Key, RawValue<BlockNumberList>) = result.unwrap();
+        let mut shard = cursor.seek(RawKey::new(sharded_key.clone()))?;
 
         // Get the highest block number that needs to be deleted for this sharded key
         let to_block = sharded_key.as_ref().highest_block_number;
 
-        match prune_shard(&mut cursor, key, raw_blocks, to_block, &key_matches)? {
-            PruneShardOutcome::Deleted => deleted += 1,
-            PruneShardOutcome::Updated => updated += 1,
-            PruneShardOutcome::Unchanged => unchanged += 1,
-        }
+        loop {
```

Comment:
nit: whenever i have nested loops, i prefer always to tag whichever has indirect control flow for clarity. e.g. `shard: loop { if ... { break 'shard }}`. up to you if you want to add this